appserver/common/container-common/src/main/java/com/sun/enterprise/container/common/impl/util/ClusteredSingletonLookupImplBase.java

```diff
@@ -60,32 +63,35 @@
     private final SingletonType singletonType;
     private final String keyPrefix;
     private final String mapKey;
-    private final String lockKey;
-    private final String sessionHzKey;
+    private final AtomicReference<String> sessionHzKey = new AtomicReference<>();
+    private final AtomicReference<String> lockKey = new AtomicReference<>();
 
     public ClusteredSingletonLookupImplBase(String componentId, SingletonType singletonType) {
         this.componentId = componentId;
         this.singletonType = singletonType;
         this.keyPrefix = makeKeyPrefix();
         this.mapKey = makeMapKey();
-        this.lockKey = makeLockKey();
-        this.sessionHzKey = makeSessionHzKey();
     }
 
-    protected String getKeyPrefix() {
+    protected final String getKeyPrefix() {
         return keyPrefix;
     }
 
-    protected String getLockKey() {
-        return lockKey;
+    protected final String getMapKey() {
+        return mapKey;
+    }
+
+    protected final String getLockKey() {
+        return lockKey.updateAndGet(v -> v != null ? v : makeLockKey());
     }
 
-    protected String getMapKey() {
-        return mapKey;
+    public final String getSessionHzKey() {
+        return sessionHzKey.updateAndGet(v -> v != null ? v : makeSessionHzKey());
     }
 
-    public String getSessionHzKey() {
-        return sessionHzKey;
+    protected final void update() {
```

Comment:
I am glad you asked :) I were describing this in _Core Server_ channel. With lombok's `@Getter` the fields were lazy initialised to a final state. That means that on first call of any of the getters the according `makeXyZ` method was computing whatever became the final value of the field. With `makeSessionHzKey` and `makeLockKey` being dependent on `getClusteredSessionKey` which reads a field that can be changed via `ClusteredSingletonLookupImpl.setClusteredSessionKey` it looks like the result of the getter should actually update when `setClusteredSessionKey` is called to change the basis on which `getClusteredSessionKey` is computed. Usage of `setClusteredSessionKey` suggests this is _dynamic_. While this is a change of behaviour to the lombok `@Getter` version I think it is what we were trying to make happen. So when `setClusteredSessionKey` is called I call the new `update` method which invalidates the `lockKey` and `sessionHzKey` used so far so that they are recomputed on next usage of the respective getter method.

Reply:
Ah I see! Could the method name be changed or a javadoc added to explain this?

Reply:
Changed it to `invalidateKeys` and added some javadoc too.